### PR TITLE
Monotonic time, fix sendfile tests, add more timing&size metrics

### PR DESCRIPTION
**Pull request by deadtrickster**
So I started with monotonic time and then changed a few other things too :D. I'll be happy to break this up to individual commits.

Timing metrics: 

```
request_start
    headers_start
    ...headers parsing...
    headers_end

    body_start
    ...body parsing...
    body_end

    user_start
    ...callback code...
    user_end

   send_start
   ...sending reply....
   send_end
request_end
```

Also added response size metrics in bytes:

```
resp_headers
resp_body  %% regualar response body size
file  %% sendfile response size
chunks %% all chunks size (wire size), not equal to Content-Length!
```

I've changed files handling - now ranges/content-size handled by Elli.

Fixed sendfile tests (looks like migration to new rebar broke them?)
Added tests for timing & size metrics.

### CODE REVIEW

File: src/elli_http.erl [hunk 1]
```diff
@@ -113,7 +115,8 @@ handle_request(S, PrevB, Opts, {Mod, Args} = Callback) ->
 
             t(request_end),
             handle_event(Mod, request_complete,
-                         [Req1, handover, [], <<>>, get_timings()], Args),
+                         [Req1, handover, [], <<>>, {get_timings(),
+                                                     get_sizes()}], Args),
```
Comment (yurrriq) [inline src/elli_http.erl:119]: We should update the docs and examples for this change.

**Comment by yurrriq** [comment]
Looks like `erlang:monotonic_time/0` is 18+. I'd be inclined to drop support for <18 rather than add a patch. @elli-lib/team?

File: test/elli_tests.erl [hunk 2]
```diff
@@ -3,56 +3,89 @@
 -include("elli.hrl").
 
 -define(I2B(I), list_to_binary(integer_to_list(I))).
+-define(I2L(I), integer_to_list(I)).
 -define(README, "README.md").
+-define(VTB(T1,T2, LB, UB),
```
Comment (yurrriq) [inline test/elli_tests.erl]: Totally pedantic, but there should be a space between `T1` and `T2`.

File: test/elli_tests.erl [hunk 3]
```diff
@@ -3,56 +3,89 @@
...
+        erlang:convert_time_unit(
+          get_timing_value(T2) -
+              get_timing_value(T1),
+          native,
+          micro_seconds) > LB andalso
+        erlang:convert_time_unit(
+          get_timing_value(T2) -
+              get_timing_value(T1),
+          native,
+          micro_seconds) < UB).
```
Comment (yurrriq) [inline test/elli_tests.erl]: Could we abstract away the `erlang:convert_time_unit(get_timing_Value(T2) - get_timing_value(T1), native, micro_seconds)`?

File: src/elli_http.erl (hunk 1 above)
Comment (yurrriq) [reply src/elli_http.erl:119]: Here's a patch to get started:

``` diff
diff --git a/src/elli_example_callback.erl b/src/elli_example_callback.erl
index b9e8cbe..6d267ec 100644
--- a/src/elli_example_callback.erl
+++ b/src/elli_example_callback.erl
@@ -233,10 +233,11 @@ chunk_loop(Ref, N) ->
 %% tree.
 %%
 %% `request_complete' fires *after* Elli has sent the response to the
-%% client. Timings contains timestamps of events like when the
+%% client. `Timings' contains timestamps of events like when the
 %% connection was accepted, when request parsing finished, when the
-%% user callback returns, etc. This allows you to collect performance
-%% statistics for monitoring your app.
+%% user callback returns, etc.
+%% TODO: Describe `Sizes' here.
+%% This allows you to collect performance statistics for monitoring your app.
 %%
 %% `request_throw', `request_error' and `request_exit' events are sent if
 %% the user callback code throws an exception, has an error or
@@ -289,7 +290,7 @@ chunk_loop(Ref, N) ->
 handle_event(elli_startup, [], _) -> ok;
 handle_event(request_complete, [_Request,
                                 _ResponseCode, _ResponseHeaders, _ResponseBody,
-                                _Timings], _) -> ok;
+                                {_Timings, _Sizes}], _) -> ok;
 handle_event(request_throw, [_Request, _Exception, _Stacktrace], _) -> ok;
 handle_event(request_error, [_Request, _Exception, _Stacktrace], _) -> ok;
 handle_event(request_exit, [_Request, _Exception, _Stacktrace], _) -> ok;
@@ -298,7 +299,7 @@ handle_event(invalid_return, [_Request, _ReturnValue], _) -> ok;

 handle_event(chunk_complete, [_Request,
                               _ResponseCode, _ResponseHeaders, _ClosingEnd,
-                              _Timings], _) -> ok;
+                              {_Timings, _Sizes}], _) -> ok;

 handle_event(request_closed, [], _) -> ok;

```

**Review by yurrriq — CHANGES_REQUESTED** [review]
We need to update the docs re: `{Timings, Sizes}` and decided whether or not to support OTP <18.

**Comment by yurrriq** [comment]
I'm not sure why [this failed](https://travis-ci.org/elli-lib/elli/jobs/161114644#L309) on Travis. Works for me locally on 19.0.5.

File: test/elli_tests.erl (hunk 2 above)
Comment (deadtrickster) [reply test/elli_tests.erl]: fixed

File: test/elli_tests.erl (hunk 3 above)
Comment (deadtrickster) [reply test/elli_tests.erl]: extracted to the function

**Comment by deadtrickster** [comment]
Personally I'd vote for dropping OTP<18.

**Comment by yurrriq** [comment]
Can you comment out `- 17.5` in `.travis.yml` then too?

File: test/elli_tests.erl [hunk 4]
```diff
@@ -3,56 +3,88 @@
 -include("elli.hrl").
 
 -define(I2B(I), list_to_binary(integer_to_list(I))).
+-define(I2L(I), integer_to_list(I)).
 -define(README, "README.md").
+-define(VTB(T1, T2, LB, UB),
+        time_diff_to_micro_seocnds(T1, T2) > LB andalso
+        time_diff_to_micro_seocnds(T1, T2) < UB).
+
+time_diff_to_micro_seocnds(T1, T2) ->
```
Comment (yurrriq) [inline test/elli_tests.erl]: Can we `s/seocnds/seconds/` here?

Comment (deadtrickster) [reply test/elli_tests.erl]: woops :-)

**Comment by deadtrickster** [comment]
Added docs, commented 17.5

**Comment by yurrriq** [comment]
Hey, can you regenerate the docs again?

I made a minor grammar tweak in https://github.com/elli-lib/elli/pull/10/commits/a072aed50a759bb23da286eaf4242ac3a02e5e42. It's cool if you want to do `rebar3 edoc` then squash/ the last two commits.

After that I think this is good to go.

**Comment by yurrriq** [comment]
Actually we should do a version bump, maybe to `1.2.5`.

**Comment by deadtrickster** [comment]
done

**Comment by yurrriq** [comment]
Awesome, thanks!

Close #7 and close #8.

**Review by yurrriq — APPROVED** [review]

**Pull request merged**